### ISSUE TITLE
Should 'SET NOT NULL' be caught?

### ISSUE THREAD

**Issue by christianbundy** [3YOURMIND/django-migration-linter#158]
Hey there. I'm looking at some behavior I don't quite understand, so I thought I'd open a ticket to learn a bit more.

Background: django-migration-linter 3.0.0, django 3.1, python 3.6, postgres 11.

# Model

## Before

```python
foo = models.OneToOneField(
    to=Foo, null=True, on_delete=models.SET_NULL, blank=True
)
```

## After

```python3
foo = models.OneToOneField(
    to=Foo, null=False, on_delete=models.DO_NOTHING, blank=True
)
```

# Migration

When I run `./manage.py makemigrations` I'm prompted with:

```
You are trying to change the nullable field 'foo' on example to non-nullable without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Ignore for now, and let me handle existing rows with NULL myself (e.g. because you added a RunPython or RunSQL operation to handle NULL values in a previous data migration)
 3) Quit, and let me add a default in models.py
```

## ✅   Option 1

When I choose 'option 1' I'm alerted to the problem.

```
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> 42
```

```python3
class Migration(migrations.Migration):

    dependencies = [
        ("example", "0001_auto"),
    ]

    operations = [
        migrations.AlterField(
            model_name="example",
            name="foo",
            field=models.OneToOneField(
                blank=True,
                default=42,
                on_delete=django.db.models.deletion.DO_NOTHING,
                to="foo.foo",
            ),
            preserve_default=False,
        ),
    ]
```

```sql
BEGIN;
--
-- Alter field foo on example
--
SET CONSTRAINTS "example_g_example_id_2fa2c87b_fk_common_oa" IMMEDIATE; ALTER TABLE "example_example" DROP CONSTRAINT "example_g_example_id_2fa2c87b_fk_foo_oa";
ALTER TABLE "api_example_example" ALTER COLUMN "foo_id" SET DEFAULT 42;
UPDATE "example_example" SET "foo_id" = 42 WHERE "foo_id" IS NULL;
ALTER TABLE "example_example" ALTER COLUMN "foo_id" SET NOT NULL;
ALTER TABLE "example_example" ADD CONSTRAINT "example_g_example_id_2fa2c87b_fk_foo_oa" FOREIGN KEY ("foo_id") REFERENCES "foo_foo" ("id") DEFERRABLE INITIALLY DEFERRED;
ALTER TABLE "example_example" ALTER COLUMN "foo_id" DROP DEFAULT;
COMMIT;
```

django-migration-linter identifies this as a problem:

```
(example, 0002_auto)... ERR
	NOT NULL constraint on columns
```

## 🐛  Option 2

When I choose option 2, django-migration-linter doesn't warn me about the problem.

```python3
class Migration(migrations.Migration):

    dependencies = [
        ("example", "0001_auto"),
    ]

    operations = [
        migrations.AlterField(
            model_name="example",
            name="foo",
            field=models.OneToOneField(
                blank=True,
                on_delete=django.db.models.deletion.DO_NOTHING,
                to="foo.foo",
            ),
        ),
    ]
```

```sql
BEGIN;
--
-- Alter field foo on example
--
SET CONSTRAINTS "example_g_example_id_2fa2c87b_fk_foo_oa" IMMEDIATE;
ALTER TABLE "example_example" DROP CONSTRAINT "example_g_example_id_2fa2c87b_fk_foo_oa";
ALTER TABLE "example_example" ALTER COLUMN "foo" SET NOT NULL;
ALTER TABLE "example_example" ADD CONSTRAINT "example_g_example_id_2fa2c87b_fk_foo_oa" FOREIGN KEY ("foo") REFERENCES "foo_foo" ("id") DEFERRABLE INITIALLY DEFERRED;
COMMIT;
```

Unfortunately django-migration-linter doesn't seem to have a problem with this:

```
(example, 0002_auto)... OK
```

Is this maybe a bug in django-migration-liniter, or am I misunderstanding what kinds of problems it catches?

**Comment by christianbundy** [comment]
If there _is_ a problem, I think it's here:

```python3
def has_not_null_column(sql_statements, **kwargs):
    # TODO: improve to detect that the same column is concerned
    ends_with_default = None
    for sql in sql_statements:
        if "SET DEFAULT" in sql:
            ends_with_default = True
        elif "DROP DEFAULT" in sql:
            ends_with_default = False
    return (
        any(
            re.search("(?<!DROP )NOT NULL", sql) and not sql.startswith("CREATE TABLE")
            for sql in sql_statements
        )
        and ends_with_default is False
    )
```

It looks like the `any()` at the end returns `True` for both, but `ends_with_default` differs. Is `ALTER COLUMN "foo" SET NOT NULL` backward-compatible if it doesn't set a default? If an older codebase tries to insert and leaves this column `null`, I'd imagine it would error.

**Comment by David-Wobrock** [comment]
Hi @christianbundy 

Thanks a lot for opening this bug. Very clear explanation, and the analysis seems to correct to me.
I'll dive into the issue and check out your PR in order to get it merged :)
